routemaster/state_machine/actions.py

```diff
@@ -100,15 +103,29 @@ def process_retries(
             conn,
         )
 
-        for label_name, metadata in relevant_labels.items():
-            label = LabelRef(name=label_name, state_machine=state_machine.name)
+    for label_name, metadata in relevant_labels.items():
+        label = LabelRef(name=label_name, state_machine=state_machine.name)
+        could_progress = False
 
-            with conn.begin():
-                _process_action_with_metadata(
+        with app.db.begin() as conn:
+            try:
+                lock_label(label, conn)
+                current_state = get_current_state(label, state_machine, conn)
+
+                if current_state != action:
+                    continue
+
+                could_progress = _process_action_with_metadata(
                     app,
                     action,
                     label,
                     metadata,
                     state_machine,
                     conn,
                 )
+            except Exception:
+                # This is allowed to error, we will retry again later.
+                pass
+
+        if could_progress:
+            process_transitions(app, label)
```

Comment:
We might want to catch and ignore `DeletedLabel` here (we're outside the lock where we last operated on the label).

Reply:
We might also want to think about what this means for the other uses of `process_transitions` too (I'm not sure they all want the same behaviour).

Reply:
Hmmmmm, I think we might want it in the other places... `process_transitions` could theoretically take minutes to process, and deleting a label in the middle is a legit requirement, and should just be handled gracefully.